lib/rubber/cloud/vagrant.rb

```diff
@@ -49,23 +47,98 @@ def describe_instances(instance_id=nil)
       def destroy_instance(instance_id)
         # If it's being run from vagrant, then 'vagrant destroy' must have been called already, so no need for us to do it.
         unless ENV.has_key?('RUN_FROM_VAGRANT')
-          cmd = "vagrant destroy #{instance_id} --force"
-          capistrano.logger.info("Destroying instance using: #{cmd}")
-          system(cmd)
+          run_vagrant_command('destroy', instance_id, '--force')
         end
       end
 
       def stop_instance(instance, force=false)
-        cmd = "vagrant suspend #{instance.instance_id}"
-        capistrano.logger.info("Stopping instance using: #{cmd}")
-        system(cmd)
+        run_vagrant_command('suspend', instance.instance_id)
       end
 
       def start_instance(instance)
-        cmd = "vagrant resume #{instance.instance_id}"
-        capistrano.logger.info("Starting existing instance using: #{cmd}")
-        system(cmd)
+       run_vagrant_command('resume', instance.instance_id)
       end
+
+      private
+
+      def run_vagrant_command(subcmd, instance_or_id = nil, args = "", return_result = false)
+        cmd = "vagrant #{subcmd}"
+
+        # Setup instance id and VAGRANT_CWD
+        if instance_or_id
```

Comment:
this function is overly complicated it should be broken into multiple functions.